File: s2hat/s2hat_example.c

#include "stdio.h"
#include "stdlib.h"
#include "math.h"
#include "time.h"
#include "sys/types.h"
#include "sys/stat.h"
#include "sys/param.h"
#include "mpi.h"

#include "s2hat.h"

/* this code computes first map2alm transform of some map, collects the alm coefficients in the memory       *
 * of the proc root and outputs them as a binary file 'test_alm.bin'. Then it performs alm2map and collects  *
 * the map in the memory of the proc root. It outputs the map as 'test_map.bin'. Then it computes the power  *
 * spectrum out of the precomputed alm coefficients. It is written as a file 'test_cl.bin'. Note that though *
 * in principle the code performs the 'full loop' map2alm followed by alm2map the outputted map will in not  *
 * agree perfectly with the input due to the bandwidth issues.                                               *
 *                                                                                                           *
 *                                                                                      - rs@apc, 2010/02/10 *
 * Works correctly for a multiple maps and any accepted number of Stokes parameters                          *
 *                                                                                                           *
 *                                                                                      - rs@apc, 2010/09/07 */

s2hat_int4 main(s2hat_int4 argc, char * argv[])
{

  s2hat_int4 pixchoice = PIXCHOICE_HEALPIX;   /* use HEALPIX pixelization */

  s2hat_int4 nside = 64;
  s2hat_int4 nmaps = 1;
  s2hat_int4 lmax = 4*nside-1;
  s2hat_int4 npix=12*nside*nside;
  s2hat_int4 root=0;

  s2hat_int4 myrank, nprocs;

  s2hat_int4 nlmax=lmax, nmmax=lmax;
  s2hat_int4 nstokes=3;  /* no of Stokes params 1, 2 or 3 */
  s2hat_int4 nspec = 1;  /* no of map spectra to be computed - check the rules in the manual */
  s2hat_int4 i, j, lout, mout;
  s2hat_int4 plms=0;     /* i.e., do not assume that plm are precomputed */
  s2hat_int4 nmvals, first_ring, last_ring, map_size;
  s2hat_int4 *mvals;
  s2hat_int8 nplm;
  s2hat_int4 one = 1, zero = 0;

  s2hat_flt8 *local_map, *map;
  s2hat_int4 nrings;
  s2hat_flt8 *local_w8ring;

  s2hat_flt8 *cls;

  FILE *fout;

  s2hat_int4 nalms;
  s2hat_dcomplex *local_alm, *alms;

  s2hat_pixeltype cpixelization;
  s2hat_pixparameters pixpar;
  s2hat_scandef cscan;

  s2hat_flt8 zbounds[2];

  /* MPI initialization */

  MPI_Init( &argc, &argv);

  MPI_Comm_rank( MPI_COMM_WORLD, &myrank);
  MPI_Comm_size( MPI_COMM_WORLD, &nprocs);

  /* define the maps parameters and data distribution */

  /* - define the pixelization structure -> cpixelization */
  pixpar.par1 = nside; pixpar.par2 = 0;
  set_pixelization( pixchoice, pixpar, &cpixelization); 

  /* - define the scan structure -> cscan : corresponding to the observed sky given by zbounds[0] >= cos theta >= zbounds[1] */

  zbounds[0] = 0.0; zbounds[1] = 0.0;
  zbounds2scan( zbounds, cpixelization, &cscan);

  /* computes the data distribution over the processors */

  get_local_data_sizes( plms, cpixelization, cscan, nlmax, nmmax, myrank, nprocs,
			&nmvals, &first_ring, &last_ring, &map_size, &nplm, root, MPI_COMM_WORLD);  

  /* number of equilatitude rings on this processor */
  nrings = last_ring-first_ring+1;

  /* m-values to be stored on this proc */
  mvals = (int *) calloc( nmvals, sizeof( int));
  find_mvalues( myrank, nprocs, nmmax, nmvals, mvals);

  /* number of alm coefficients stored on this proc */
  nalms = (nlmax+1)*nmvals*nstokes;

  /* to store part of the map */
  local_map = (s2hat_flt8 *)calloc( nmaps*map_size*nstokes, sizeof( s2hat_flt8));

  /* to assign some arbitrary values to the map NB. these depend on a number of pixs used */
  for( j=0; j<nmaps; j++) for( i=0; i<map_size*nstokes; i++) local_map[i+j*nstokes*map_size] = 1.0*i;

  /* quadrature weights */
  local_w8ring = (s2hat_flt8 *)calloc( nrings*nstokes, sizeof( s2hat_flt8));

  /* i.e. use no weighting */
  for( i=0; i<nrings*nstokes; i++) local_w8ring[i] = 1.0;

  local_alm = (s2hat_dcomplex *) calloc( nalms*nstokes, sizeof( s2hat_dcomplex));

  /* calculate the alm coefficients */

  s2hat_map2alm( plms, cpixelization, cscan, nlmax, nmmax, nmvals, mvals,
	         nmaps, nstokes, first_ring, last_ring,
	         local_w8ring, map_size, local_map, nlmax,
	         local_alm, nplm, NULL, nprocs, myrank, MPI_COMM_WORLD);

  free( local_w8ring);

  /* gather all computed alm in the memory of the proc 'root' */

  if( myrank == root)
  {
     alms = (s2hat_dcomplex *)calloc( nstokes*(nlmax+1)*(nmmax+1), sizeof( s2hat_dcomplex));
     fout = fopen( "test_alm.bin","w"); fclose( fout);
  }

  /* do a loop over maps */
  for( i=0; i<nmaps; i++)
  {
     collect_alms( nlmax, nmmax, nmaps, i, nstokes, nmvals, mvals, nlmax, local_alm, alms, myrank, nprocs, root, MPI_COMM_WORLD);

     if( myrank == root) 
     {
        fout = fopen( "test_alm.bin", "a"); fwrite( alms, sizeof( s2hat_dcomplex), nstokes*(nmmax+1)*(nlmax+1), fout); fclose( fout);
     }
  }

  if( myrank == root) free( alms);

  /* go back from alms to map */

  s2hat_alm2map( plms, cpixelization, cscan, nlmax, nmmax, nmvals, mvals, nmaps, nstokes, first_ring, last_ring,
		 map_size, local_map, nlmax, local_alm, nplm, NULL, nprocs, myrank, MPI_COMM_WORLD);

  /* gather complete map in the memory of the proc root */

  if( myrank == root)
  {
      map = (s2hat_flt8 *)calloc( nstokes*npix, sizeof( s2hat_flt8));
      fout = fopen( "test_map.bin","w"); fclose( fout);
  }

  /* do the loop over maps */
  for( i=0; i<nmaps; i++)
  {
     collect_map(cpixelization,nmaps,i,nstokes,map,first_ring,last_ring,map_size,local_map,myrank,nprocs,root,MPI_COMM_WORLD);

     if( myrank == root)
     {
         fout=fopen( "test_map.bin","a"); fwrite( map, sizeof( double), npix*nstokes, fout); fclose( fout);
     }
  }

  if( myrank == root) free( map);

  free( local_map);

  /* and compute (auto)spectra just for the first map */

  cls = (s2hat_flt8 *)calloc( (nlmax+1)*nspec, sizeof( s2hat_flt8));
  collect_cls( nmaps, zero, nstokes, nlmax, nmvals, mvals, nlmax, local_alm, 
               nspec, cls, myrank, nprocs, root, MPI_COMM_WORLD);

  free( local_alm);

  if( myrank == root) 
  {
     fout = fopen( "test_cl.bin", "w"); fwrite( cls, sizeof( double), (nlmax+1)*nspec, fout); fclose( fout);
  }

  free( cls);

  /* conclude */

  destroy_scan( cscan);
  destroy_pixelization( cpixelization);

  /* and finish up */

  MPI_Finalize();
  return( 0);

}

File: s2hat/s2hat_example.f90

program test

  use s2hat_defs
  use s2hat_toolbox_mod
  use s2hat_pixelization
  use s2hat_map2alm_mod
  use s2hat_alm2map_mod

  implicit none

  ! include "mpif.h"

  integer(s2hat_i4b) :: myid, ierr, numprocs, root, comm
  integer(s2hat_i4b) :: nside, npix, lmax, nmaps, i, j, l, m, nstokes
  integer(s2hat_i4b) :: nmmax, nmvals, first_ring, last_ring, map_size, lda, nrings
  integer(s2hat_i4b) :: nringsall, precompute_plms, mapnum, nspec
  integer(s2hat_i8b) :: nplm
  character(len=128) :: param
  real(s2hat_dp)     :: t1, t2, t3, t4
  real(s2hat_dp),     pointer,     dimension(:,:)   :: local_w8ring, local_plms, map, mapout, w8ring
  real(s2hat_dp),     pointer,     dimension(:,:)   :: cls  
  real(s2hat_dp),     pointer,     dimension(:,:,:) :: local_map
  complex(s2hat_dpc), pointer,     dimension(:,:,:) :: alms
  complex(s2hat_dpc), pointer,     dimension(:,:,:,:) :: local_alms
  integer(s2hat_i4b), dimension(MPI_STATUS_SIZE)    :: status
  integer(s2hat_i4b), pointer,     dimension(:)     :: mask, mvals
  real(s2hat_dp),                  dimension(2)     :: zbounds
  type(s2hat_pixparameters) :: pixpars
  type(s2hat_pixeltype)     :: pixelization
  type(s2hat_scandef)       :: scan

  call cpu_time(t1)

  call mpi_init(ierr)
  call mpi_comm_rank(MPI_COMM_WORLD, myid, ierr)
  call mpi_comm_size(MPI_COMM_WORLD, numprocs, ierr)
  root = 0
  comm = MPI_COMM_WORLD

  nside = 128
  
  lmax            = 2*nside
  npix            = 12*nside**2
  nmaps           = 1
  nstokes         = 3
  lda             = lmax         ! s2hat convention for alm matrix arrangement
  ! lda             = nstokes    ! that would be HEALPix like
  precompute_plms = 0            ! Plms are not to be precomputed
  mapnum          = 1            ! # of maps to be processed simultanously
  nmmax           = lmax

  zbounds(:) = 0.d0              ! full map to be processed - no equatorial cut

  pixpars%par1 = nside           ! for HEALPIX the second field of pixpars is irrelevant

  call set_pixelization(PIXCHOICE_HEALPIX, pixpars, pixelization)  ! precomputes the pixelization structure assuming the HEALPIX pixelization
  
  nringsall = pixelization%nringsall  ! # of rings in the Northern hemisphere plus equator

  call zbounds2scan(zbounds, pixelization, scan)  ! computes the scan structure given 'zbounds'

  ! computes the object sizes needed for memory allocation etc
  call get_local_data_sizes(precompute_plms, pixelization, scan, lmax, nmmax, myid, numprocs, nmvals, &
          & first_ring, last_ring, map_size, nplm, root, comm)

  ! allocate storage for the values of the m numbers processed on this proc
  allocate(mvals(nmvals))

  call find_mvalues(myid, numprocs, nmmax, nmvals, mvals)  ! define the m-values subset

  nrings = last_ring-first_ring+1 ! # of equatorial rings o be processed by this proc

  allocate(local_w8ring(nrings,nstokes))    ! to store the ring quadrature weights - one per ring
  allocate(local_map(0:map_size-1,nstokes,nmaps))    ! to store the submap processed by this proc

  local_w8ring(1:nrings,1:nstokes) = 1.d0   ! i.e. no ringw weighting

  do i = 0, map_size-1
     local_map(i,1:nstokes,1:nmaps) = pixelization%fpix(first_ring)+i   ! fake map
  end do

  if( lda == nstokes) then
      allocate(local_alms(nstokes,0:lmax,0:nmvals-1,nmaps))    ! healpix like arrangement
  else
      allocate(local_alms(0:lmax,0:nmvals-1,nstokes,nmaps))    ! s2hat native arrangement
  end if     

  local_alms = 0.d0

  ! map-to-alm transform

  call s2hat_map2alm(0, pixelization, scan, lmax, nmmax, nmvals, mvals, nmaps, nstokes, &
          & first_ring, last_ring, local_w8ring, map_size, local_map, lda, local_alms, &
          & nplm, local_plms, numprocs, myid, comm)

  nspec = 4  ! get TT, EE, BB, TE spectra only

  if( myid == root) allocate( cls(0:lmax,1:nspec))                     ! only on the proc 'root'

  call collect_cls( nmaps, 0, nstokes, lmax, nmvals, mvals, lmax, local_alms,  &
                  & nspec, cls, myid, numprocs, root, comm)

  ! alm-to-map
     
  call s2hat_alm2map(precompute_plms, pixelization, scan, lmax, nmmax, nmvals, &
          & mvals, nmaps, nstokes, first_ring, last_ring, map_size, local_map, lda, local_alms, &
          & nplm, local_plms, numprocs, myid, comm)

  deallocate( local_map)
  deallocate( local_alms)
  deallocate( local_w8ring)
  deallocate( mvals)

  if( myid == root) deallocate( cls)

  ! And exit
  call mpi_finalize(ierr)

end program test
